flight/PiOS/STM32F4xx/pios_usart.c

```diff
@@ -215,13 +221,50 @@ int32_t PIOS_USART_Init(uintptr_t * usart_id, const struct pios_usart_cfg * cfg,
 		break;
 	}
 	NVIC_Init((NVIC_InitTypeDef *)&(usart_dev->cfg->irq.init));
-	USART_ITConfig(usart_dev->cfg->regs, USART_IT_RXNE, ENABLE);
-	USART_ITConfig(usart_dev->cfg->regs, USART_IT_TXE,  ENABLE);
 
 	// FIXME XXX Clear / reset uart here - sends NUL char else
 
+	/* Check DMA transmit configuration and prepare DMA. */
+	if (cfg->dma_send) {
+		usart_dev->dma_send_buffer = (uint32_t)PIOS_malloc(PIOS_USART_DMA_SEND_BUFFER_SZ);
```

Comment:
There's no sane zero-copy / way to use the PIOS_Com buffers?

Reply:
Can probably do something with circ_queue.

Reply:
Yah, circ_queue was written with DMA in mind (though I can't promise perfectly so, as I've not used it this way yet).

You can get a bunch of the buffer space, for receive, and fill it and then `circ_queue_advance_write_multi`.  Similarly you can get a block of read data but keep it around until released with `circ_queue_read_completed_multi`.

The big caveats are: A) You don't want to keep too much of the transmit data around after it's been sent, clogging up the buffer... so you don't want to say, DMA 300 bytes and then only release them from the buffer all at once when the DMA is completed.  B) When's the right time to make data available to receive?  You really need per-byte interrupts still in case someone is waiting for the last byte, which means the interrupt loading isn't improved much on receive.

Might be worth skipping receive DMA even, for this reason.

Reply:
Receive DMA would be interesting to axe most of the Crossfire interrupts :P

Regarding circ_queue, I'll probably have to introduce two new callbacks for this. The current ones just copy things into a buffer you need to supply.

Reply:
As far as receive goes, it uses idle detection available on the F4. So when receive goes idle, even after the last byte, the USARTs IRQ handler stops DMA, which triggers the DMA handler and empties whatever it got into PIOS_Com. Might be a bit heavy handed for straddling bytes, but generally it improves things, see the 1.5% CPU savings mentioned earlier.

I'll probably add a way to hint that we don't want DMA, so depending on the device type, PIOS HAL can request to not use it.

Reply:
(Hand over bytes either when DMA ends OR idle).

Reply:
Ah shit this may be race-y.  Because we could be idle but the last character may not have been DMA'd yet.